### PR TITLE
Feature/smar 121+smar 136

### PR DESCRIPTION
**Pull request by SwiftXLight**
**Description changes:** 
- Ability to create appointment
- Ability to find doctors via availability
- Ability to get all appointments for exact doctor via id
- Ability to get all appointments for exact patient via id
- Ability to get list of patients, with which doctor has appointments
- Ability to find doctors with some specialization
- Rewrite availability as separate table (it was json in doctor)

[link to Jira ticket 121](https://smartdc23.atlassian.net/browse/SMAR-121?atlOrigin=eyJpIjoiMDE3NjQ0YThlZGM5NDk5NmE5NWUzMzk2MzQzN2M4YTgiLCJwIjoiaiJ9)
[link to Jira ticket 136](https://smartdc23.atlassian.net/browse/SMAR-136?atlOrigin=eyJpIjoiOTNmODZhNzJkYTI2NDU5Nzg0N2JkMjRmZWU2MmU4YjgiLCJwIjoiaiJ9)

**PR checklist:**
1.1 Common
1.1.1 types for input and output params
1.1.2 no any, should be also added to eslint rules
1.1.3 try/catch for any async function
1.1.4 no magic numbers
1.1.5 compare only with constants not with strings
(instead of user === ‘admin’, better user === roles.admin)
1.1.6 no ternary operator inside ternary operator (bad example: question ? first:
second ? third: fourth)
1.1.7 avoid similar types with duplicating by generics
1.1.8 no console.log in pr
1.1.9 .env file should be in .gitignore
1.1.10 delete commented code if it's not part of documentation
1.1.11 no links in the code, env links should be in env file (for example: server url),
constant links (for example default avatar URL) should be in constant file
1.1.12 constants and function names should be lowercase.
1.1.13 no dates format in the code (like "dd/MM/yyyy”), move it in global constant
variable
1.1.14 import rules. imports should come in a specific order: node modules, absolute
path, relative path
1.1.15 strings should be in the constant variable. Example: instead of ‘15 3 * * *’,
const EACH_DAY_15h_15min = ‘15 3 * * *’

1.2 Backend
1.2.1 swagger for each endpoint
1.2.2 less requests to db, all data should be taken with one query
1.2.3 `public` in method use only is use function externally
1.2.4 use ConfigService instead of `process.env`

Appointment table:
image
image

Create new appointment:
image

Get appointments, which are connected with doctor id:
image

Get appointments, which are connected with patient id:
image

Get list of patients, which have appointments with doctor:
image

Get all doctors with some specialization:
image

Get all doctors which availability match with input:
image

Availability table:
image

### CODE REVIEW

File: src/modules/appointment/appointment.service.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        .into(Appointment)
+        .values(newAppointment)
+        .execute()
+        .then((result) => {
+          const { generatedMaps } = result;
+          const [generatedMap] = generatedMaps;
+          const { id } = generatedMap;
+
+          return { ...newAppointment, id };
+        });
```
Comment (lyudad) [inline src/modules/appointment/appointment.service.ts]: find other way, make 3 select query to make one insert too expencive

File: src/modules/appointment/appointment.service.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        .leftJoinAndSelect('appointment.patient', 'patient')
+        .where(
+          'appointment.localDoctorId = :id OR appointment.remoteDoctorId = :id',
+          { id: doctor.id },
+        )
+        .getMany();
+
+      const patients: Patient[] = appointments.map(
+        (appointment) => appointment.patient,
+      );
```
Comment (lyudad) [inline src/modules/appointment/appointment.service.ts]: can we find other way, map through each appointment is expencive

File: src/modules/availability/availability.service.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+          `Availability with uuid ${availabilityId} not found`,
+        );
+      }
+    } catch (err) {
+      throw new HttpException(`${err}`, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  async findDoctorsByAvailability(
+    start: string,
```
Comment (lyudad) [inline src/modules/availability/availability.service.ts]: make better naming: startDate, startTime

File: src/modules/doctor/entity/doctor.entity.ts [hunk 4]
```diff
@@ -1,12 +1,6 @@
...
-import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
-
-export interface Availability {
-  uuid: string;
-  title: string;
-  start: Date;
-  end: Date;
-}
+import { Entity, Column, PrimaryGeneratedColumn, ManyToOne } from 'typeorm';
+import Availability from '../../availability/entity/availability.entity';
```
Comment (lyudad) [inline src/modules/doctor/entity/doctor.entity.ts]: absolute path

Comment (SwiftXLight) [reply src/modules/doctor/entity/doctor.entity.ts]: fixed

File: src/modules/availability/availability.service.ts (hunk 3 above)
Comment (SwiftXLight) [reply src/modules/availability/availability.service.ts]: fixed

File: src/modules/appointment/appointment.service.ts (hunk 1 above)
Comment (SwiftXLight) [reply src/modules/appointment/appointment.service.ts]: rewrited with single query

File: src/modules/appointment/appointment.service.ts (hunk 2 above)
Comment (SwiftXLight) [reply src/modules/appointment/appointment.service.ts]: rewrited with subQuery

File: src/modules/appointment/appointment.service.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    appointment: CreateAppointmentDto,
+  ): Promise<Appointment> {
+    try {
+      const { localDoctorId, remoteDoctorId, patientId, ...rest } = appointment;
+
+      const [patient, localDoctor, remoteDoctor] = await Promise.all([
+        this.patientService.getPatientById(patientId),
+        this.doctorService.getDoctorByID(localDoctorId),
+        this.doctorService.getDoctorByID(remoteDoctorId),
+      ]);
```
Comment (lyudad) [inline src/modules/appointment/appointment.service.ts:28]: why we need 3 query, to. make insert to Appointment we need id,not full object
find other way

Comment (SwiftXLight) [reply src/modules/appointment/appointment.service.ts:28]: rewrited and inserted only ids on appointment

File: src/modules/appointment/appointment.controller.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    @Body() appointment: CreateAppointmentDto,
+  ): Promise<Appointment> {
+    return this.appointmentService.createAppointment(appointment);
+  }
+
+  @ApiOperation({ summary: "Find doctor's appointments" })
+  @ApiResponse({ status: 200, type: [Appointment] })
+  @Get('/doctor/:id')
+  async getAppointmentsByDoctorId(
+    @Param('id') id: number,
```
Comment (lyudad) [inline src/modules/appointment/appointment.controller.ts:27]: I think doctor Id could be taken from token, no?

**Review by lyudad — APPROVED** [review]

**Pull request merged**